src/rng.rs

```diff
@@ -0,0 +1,115 @@
+use core::cmp;
+use core::mem;
+
+use crate::hal::blocking::rng;
+use crate::stm32::RNG;
+use crate::stm32;
+
+#[derive(Debug)]
+pub enum ErrorKind {
+    ClockError,
+    SeedError,
+}
+
+
+pub trait RngExt {
+    fn constrain(self) -> Rng;
+}
+
+impl RngExt for RNG {
+
+    /// Enable RNG_CLK and the RNG peripheral.
+    /// Note that clocks must already be configured
+    /// such that RNG_CLK is not less than 1/16 HCLK,
+    /// otherwise all reads of the RNG will return
+    /// a ClockError (CECS error)
+    fn constrain(self) -> Rng {
+        let rcc = unsafe { &*stm32::RCC::ptr() };
+
+        cortex_m::interrupt::free(|_| {
+            // enable RNG_CLK (peripheral clock)
+            rcc.ahb2enr.modify(|_, w| w.rngen().enabled());
+
+            // give RNG_CLK time to start
+            let _ = rcc.ahb2enr.read().rngen().is_enabled();
+
+            // reset the RNG
+            rcc.ahb2rstr.modify(|_, w| w.rngrst().set_bit());
+            rcc.ahb2rstr.modify(|_, w| w.rngrst().clear_bit());
+
+            // enable the RNG peripheral
+            self.cr.modify(|_, w| w.rngen().set_bit());
+        });
+
+        Rng { rb: self }
+    }
+}
+
+pub trait RngCore<W> {
+    fn gen(&mut self) -> Result<W, ErrorKind>;
+    fn fill(&mut self, dest: &mut [W]) -> Result<(), ErrorKind>;
+}
+
+pub struct Rng {
+    rb: RNG,
+}
+
+impl Rng {
+    /// Returns 32 bits of random data, or error
+    pub fn next(&mut self) -> Result<u32, ErrorKind> {
+        loop {
+            let status = self.rb.sr.read();
+            if status.cecs().bit() {
+                return Err(ErrorKind::ClockError);
+            }
+            if status.secs().bit() {
+                return Err(ErrorKind::SeedError);
+            }
+            if status.drdy().bit() {
+                return Ok(self.rb.dr.read().rndata().bits());
+            }
+        }
+    }
+
+    pub fn release(self) -> RNG {
+        self.rb
+    }
+}
+
+impl rng::Read for Rng {
+    type Error = ErrorKind;
+
+    fn read(&mut self, buffer: &mut [u8]) -> Result<(), Self::Error> {
+        self.fill(buffer)
+    }
+}
+
+macro_rules! rng_core {
+    ($($type:ty),+) => {
+        $(
+            impl RngCore<$type> for Rng {
+                /// Returns a single element with random value, or error
+                fn gen(&mut self) -> Result<$type, ErrorKind> {
+                    let val = self.next()?;
+                    Ok(val as $type)
+                }
+
+                /// Fills buffer with random values, or return error
+                fn fill(&mut self, buffer: &mut [$type]) -> Result<(), ErrorKind> {
+                    const BATCH_SIZE: usize = 4 / mem::size_of::<$type>();
+                    let mut i = 0_usize;
+                    while i < buffer.len() {
+                        let random_word = self.next()?;
+                        let bytes: [$type; BATCH_SIZE] = unsafe { mem::transmute(random_word) };
```

Comment:
Not too hot about the transmute here. Probably should find a safer way to do it.

Reply:
https://doc.rust-lang.org/beta/std/primitive.u64.html#method.to_ne_bytes

So:
```suggestion
                        let bytes = random_word.to_ne_bytes();
```

Reply:
Maybe a bit more complex, but `from_ne_bytes()` and `to_ne_bytes()` looks like good tools.